source/common/tracing/zipkin/zipkin_core_types.h

```diff
@@ -0,0 +1,586 @@
+   */
+  void setServiceName(const std::string& service_name) { service_name_ = service_name; }
+
+  /**
+   * Serializes the endpoint as a Zipkin-compliant JSON representation as a string.
+   *
+   * @return a stringified JSON.
+   */
+  const std::string& toJson() override;
+
+private:
+  std::string ipv4_;
+  uint16_t port_;
+  std::string service_name_;
+  std::string ipv6_;
+
+  bool isset_ipv6_;
+};
+
+/**
+ * Represents a Zipkin basic annotation. This class is based on Zipkin's Thrift definition of
+ * an annotation.
+ */
+class Annotation : public ZipkinBase {
+public:
+  /**
+   * Copy constructor.
+   */
+  Annotation(const Annotation&);
+
+  /**
+   * Assignment operator.
+   */
+  Annotation& operator=(const Annotation&);
+
+  /**
+   * Default constructor. Creates an empty annotation.
+   */
+  Annotation() : timestamp_(0), value_(), isset_endpoint_(false) {}
+
+  /**
+   * Constructor that creates an annotation based on the given parameters.
+   *
+   * @param timestamp A 64-bit integer containing the annotation timestasmp attribute.
+   * @param value A string containing the annotation's value attribute. Valid values
+   * appear on ZipkinCoreConstants. The most commonly used values are "cs", "cr", "ss" and "sr".
+   * @param endpoint The endpoint object representing the annotation's endpoint attribute.
+   */
+  Annotation(uint64_t timestamp, const std::string value, Endpoint& endpoint)
+      : timestamp_(timestamp), value_(value), endpoint_(endpoint), isset_endpoint_(true) {}
+
+  /**
+   * @return the annotation's endpoint attribute.
+   */
+  const Endpoint& endpoint() const { return endpoint_; }
+
+  /**
+   * Sets the annotation's endpoint attribute (copy semantics).
+   */
+  void setEndpoint(const Endpoint& endpoint) {
+    endpoint_ = endpoint;
+    isset_endpoint_ = true;
+  }
+
+  /**
+   * Sets the annotation's endpoint attribute (move semantics).
+   */
+  void setEndpoint(const Endpoint&& endpoint) {
+    endpoint_ = endpoint;
+    isset_endpoint_ = true;
+  }
+
+  /**
+   * @return the annotation's timestamp attribute.
+   */
+  uint64_t timestamp() const { return timestamp_; }
+
+  /**
+   * Sets the annotation's timestamp attribute.
+   */
+  void setTimestamp(uint64_t timestamp) { timestamp_ = timestamp; }
+
+  /**
+   * return the annotation's value attribute.
+   */
+  const std::string& value() const { return value_; }
+
+  /**
+   * Sets the annotation's value attribute.
+   */
+  void setValue(const std::string& value) { value_ = value; }
+
+  /**
+   * @return true if the endpoint attribute is set, or false otherwise.
+   */
+  bool isSetEndpoint() const { return isset_endpoint_; }
+
+  /**
+   * Serializes the annotation as a Zipkin-compliant JSON representation as a string.
+   *
+   * @return a stringified JSON.
+   */
+  const std::string& toJson() override;
+
+private:
+  uint64_t timestamp_;
+  std::string value_;
+  Endpoint endpoint_;
+
+  bool isset_endpoint_;
```

Comment:
For all of the isset_ stuff it would better if you just did Optional<Endpoint> etc. (Applies elsewhere)